### PR TITLE
TEAM/MRAIL: initial implementation

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**

### CODE REVIEW

File: src/core/xccl_team_lib.h [hunk 1]
```diff
@@ -124,7 +124,7 @@ xccl_local_proc_info_t* xccl_local_process_info();
 
 #define XCCL_CONTEXT_SUPER_INIT(_ctx, _lib, _params) do {                 \
         (_ctx).lib = (_lib);                                              \
-        memcpy(&((_ctx).params), (_params), sizeof(xccl_team_params_t));  \
```
Comment (vspetrov) [inline src/core/xccl_team_lib.h:127]: was there a memory corruption before this fix?

File: src/team_lib/multirail/xccl_mrail_lib.c [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+    }
+
+    return XCCL_OK;
+free_teams:
+    team->n_teams -= 1;
+    for(; team->n_teams >= 0; team->n_teams--) {
+        xccl_team_destroy(team->teams[team->n_teams]);
+    }
+    free(team);
+}
```
Comment (vspetrov) [inline src/team_lib/multirail/xccl_mrail_lib.c:268]: missing return status

File: src/team_lib/multirail/xccl_mrail_lib.c [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+
+xccl_status_t xccl_mrail_team_destroy(xccl_tl_team_t *team)
+{
+    xccl_mrail_team_t *tl = ucs_derived_of(team, xccl_mrail_team_t);
+
+    xccl_mrail_debug("Entering team destroy");
```
Comment (vspetrov) [inline src/team_lib/multirail/xccl_mrail_lib.c]: why tl->teams[i] for i in {0,tl->n_teams} are not destroyed here?

File: src/team_lib/multirail/xccl_mrail_lib.c [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+    free(tl);
+    xccl_mrail_debug("Exiting team destroy");
```
Comment (vspetrov) [inline src/team_lib/multirail/xccl_mrail_lib.c]: i would leave only one debug message

File: src/team_lib/multirail/xccl_mrail_lib.c [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+
+    return XCCL_OK;
+}
+
+xccl_status_t xccl_mrail_context_progress(xccl_tl_context_t *team_context)
```
Comment (vspetrov) [inline src/team_lib/multirail/xccl_mrail_lib.c:284]: need to progress all contexts here.

File: src/team_lib/multirail/xccl_mrail_lib.c [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+    req = (xccl_mrail_coll_req_t*)malloc(sizeof(xccl_mrail_coll_req_t));
+    
+    if (req == NULL) {
+        return XCCL_ERR_NO_MEMORY;
+    }
+
+    if (coll_args->buffer_info.len < team->n_teams) {
+        n_reqs  = 1;
+        buf_len = 0;
+    }
```
Comment (vspetrov) [inline src/team_lib/multirail/xccl_mrail_lib.c]: why "else" on a new line?

File: src/team_lib/multirail/xccl_mrail_lib.c [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+    memset(req->reqs, 0, n_reqs * sizeof(xccl_coll_req_h));
+    team_args = *coll_args;
+
+    for(req->n_reqs = 0; req->n_reqs < n_reqs; req->n_reqs++) {
+        len = buf_len;
+        if (req->n_reqs == n_reqs - 1) {
+            len = last_buf_len;
+        }
+        xccl_mrail_debug("len %d \n", len);
+        team_args.buffer_info.dst_buffer = coll_args->buffer_info.dst_buffer +
```
Comment (vspetrov) [inline src/team_lib/multirail/xccl_mrail_lib.c]: ptr arithmetics on void * is a bad practice (https://stackoverflow.com/questions/3523145/pointer-arithmetic-for-void-pointer-in-c). Please use smth like:
(void*)((ptrdiff_t)coll_args->buffer_info.dst_buffer + len*req->nreqs)

**Review by vspetrov — CHANGES_REQUESTED** [review]

File: src/core/xccl_team_lib.h (hunk 1 above)
Comment (Sergei-Lebedev) [reply src/core/xccl_team_lib.h:127]: yes

**Review by vspetrov — APPROVED** [review]

**Pull request merged**